wandb/sdk/launch/runner/gcp_vertex.py

```diff
@@ -177,59 +160,118 @@ def run(
+    except Exception as e:
+        wandb.termwarn(f"{LOG_PREFIX}Unable to read gcloud config for {config}")
+        wandb.termwarn(f"{LOG_PREFIX}Error: {e}")
+
+    # If we can't read the config, return an empty dict
+    return {"properties": {}}
+
+
+def resolve_gcp_region(resource_args, gcp_config, registry_config):
+    """Resolve the GCP region from resource args, gcp config, and registry config.
+
+    Args:
+        resource_args: The resource args passed to the backend.
+        gcp_config: The gcloud config pulled from the local environment.
+        registry_config: The agent level registry config.
+
+    Returns:
+        The GCP region to use for compute and storage.
+
+    Raises:
+        LaunchError: If the region cannot be resolved.
+    """
+    registry_region = registry_config.get("region")
+    if registry_region:
+        # TODO: validate the region is valid
+        return registry_region
+    gcp_zone = resource_args.get("gcp_region") or gcp_config["properties"].get(
+        "compute", {}
+    ).get("zone")
+    if not gcp_zone:
+        raise LaunchError(
+            "Vertex requires a region for compute. Specify a region under key gcp_region."
+        )
+    gcp_region = "-".join(gcp_zone.split("-")[:2])
+    return gcp_region
+
+
+def resolve_artifact_repo(
+    resource_args: dict, registry_config: dict, gcp_project: str, gcp_region: str
+) -> str:
+    """Resolve the Artifact Registry repo from resource args and registry config.
+
+    Args:
+        resource_args: The resource args passed to the backend.
+        registry_config: The registry config from gcloud.
+        gcp_project: The gcp project, in case the image uri needs to be inferred.
+        gcp_region: The default region in case the docker host needs to be inferred.
+
+    Returns:
+        The resolved repo as a str.
+
+    Raises:
+        LaunchError: If repo is not set in either resource args or registry config.
+    """
+
+    # If a full repo uri is specified, use that
+    registry_repo = registry_config.get("url")
+    if registry_repo:
+        # TODO: Do some kind of validation that this repo is valid
+        return registry_repo
+
+    # Otherwise, use the repo specified in resource args
+    docker_host = resolve_docker_host(resource_args, gcp_region)
+    repo_name = resource_args.get("artifact_repo")
+    if not repo_name:
+        raise LaunchError(
+            "Vertex requires that you specify an Artifact Registry repository. "
+            "Please specify a repo in your resource args under key artifact_repo or "
+            "in your launch agent registry config under key url."
+        )
+    repository = "/".join([docker_host, gcp_project, repo_name])
+    # TODO: verify that this repo is valid
+    return repository
+
+
+def resolve_docker_host(resource_args, gcp_region):
+    """Resolve the docker host from resource args and gcloud configuration."""
+    resource_host = resource_args.get("docker_host")
```

Comment:
nit: wouldn't 
```
return resource_args.get("docker_host") or f"{gcp_region}-docker.pkg.dev"
```
suffice?

Reply:
Yeah... my reasoning in breaking this out was that these docker host urls might be entered inconsistently and require some more processing that we can test. So I just decided to break it out now